src/nanotron/data/dataset_builder.py

```diff
@@ -0,0 +1,214 @@
+import math
+from typing import Any, List, Type, Union
+
+import numpy
+import torch
+
+from nanotron import logging
+from nanotron.data.blended_nanoset import BlendedNanoset
+from nanotron.data.indexed_dataset import MMapIndexedDataset
+from nanotron.data.nanoset import Nanoset
+from nanotron.data.nanoset_configs import NanosetConfig
+from nanotron.data.utils import Split, normalize
+
+logger = logging.get_logger(__name__)
+
+DistributedDataset = Union[Nanoset, MMapIndexedDataset, BlendedNanoset]
+
+
+class NanosetBuilder(object):
+    """Builder class for the Nanoset classes
+
+    Args:
+
+        config (NanosetConfig): The config object which informs dataset creation
+    """
+
+    def __init__(
+        self,
+        config: NanosetConfig,
+    ):
+        self.config = config
+
+    def build(self) -> List[Union[BlendedNanoset, Nanoset]]:
+        """Build all dataset splits according to the provided data_path(s)
+
+        This method is distributed-aware and must be called on all ranks.
+
+        The dataset splits returned can vary according to the config. Supply config.data_path and
+        config.split to build BlendedNanoset and/or Nanoset splits from the same
+        distribution.
+
+        Returns:
+            List[Union[BlendedNanoset, Nanoset]]: A list of either Nanoset or BlendedNanoset per split
+        """
+
+        data_path = self.config.data_path
+        split = self.config.split_vector
+
+        # Single Nanoset
+        if isinstance(data_path, str):
+            return self.build_nanoset_dataset_splits(data_path, split, self.config.split_num_samples)
+
+        # Blended Nanoset
+        prefix_per_dataset = list(data_path.keys())
+        weight_per_dataset = normalize(list(data_path.values()))
+
+        # NOTE: Use 0.5% target margin to ensure that that we do not exhaust the indices of any dataset in the Blend
+        sizes_per_dataset = [
```

Comment:
We should add some logging to track the number of samples per dataset based on the weights and the total number of samples after weighting?

Reply:
[Done!](https://github.com/TJ-Solergibert/nanotron/blob/531f802a215a605a9d75f770ea83ba9da600b52b/src/nanotron/data/dataset_builder.py#L93)